thor/src/protocols/splice.rs

```diff
@@ -288,32 +387,52 @@ pub(crate) struct State3 {
     y_self: PublishingKeyPair,
     Y_other: PublishingPublicKey,
     previous_TX_f_output_descriptor: Descriptor<bitcoin::PublicKey>,
-    TX_f: FundingTransaction,
-    sig_TX_f: Signature,
+    splice_transaction: SpliceTransaction,
+    splice_transaction_signature: Signature,
     TX_c: CommitTransaction,
     signed_TX_s: SplitTransaction,
     encsig_TX_c_self: EncryptedSignature,
     encsig_TX_c_other: EncryptedSignature,
+    signed_splice_transaction: Option<PartiallySignedTransaction>,
 }
 
 impl State3 {
     pub async fn next_message(&self) -> anyhow::Result<Message3> {
         Ok(Message3 {
-            sig_TX_f: self.sig_TX_f.clone(),
+            splice_transaction_signature: self.splice_transaction_signature.clone(),
+            signed_splice_transaction: self.signed_splice_transaction.clone(),
         })
     }
 
     /// Returns the Channel and the transaction to broadcast.
-    pub fn receive(
+    pub async fn receive(
         self,
         Message3 {
-            sig_TX_f: sig_TX_f_other,
+            splice_transaction_signature: splice_transaction_signature_other,
+            signed_splice_transaction: signed_splice_transaction_other,
         }: Message3,
+        wallet: &impl SignFundingPSBT,
     ) -> anyhow::Result<(Channel, Transaction)> {
-        let signed_TX_f = self.TX_f.clone().add_signatures(
+        // If the other party sent a splice-in signed TX_f, use it, otherwise, use our
+        // unsigned TX_f
+        let splice_transaction = match signed_splice_transaction_other {
+            Some(signed_splice_transaction_other) => signed_splice_transaction_other,
+            None => self.splice_transaction.clone().into_psbt()?,
+        };
+
+        // If we have a splice-in input, we need to sign it, otherwise, use the previous
+        // TX_f
+        let splice_transaction = match self.signed_splice_transaction {
+            Some(_) => wallet.sign_funding_psbt(splice_transaction).await?,
```

Comment:
We should double check the `splice_transaction` if it fits with the inputs and outputs we want to provide, i.e. compare against `self.splice_transaction`.

Reply:
I added a todo for now as I have a lot of rebase to sort out.